File: tabulation/golang/02_grid_traveler.go
// Grid traveler.

// Implementation of a solution to the grid traveler problem using tabulation.

// Say that you are a traveler on a 2D grid. You begin in the top-left corner and your goal
// is to travel to the bottom-right corner. You may only move down or right.

// In how many ways can you travel to the goal on a grid with dimensions m * n?
package main

import "fmt"

// Given the number of lines and columns, return the number of ways the traveler can
// travel to the goal.
func GridTraveler(m int, n int) int {
	var table [][]int

	for i := 0; i <= m; i++ {
		var row []int
		for j := 0; j <= n; j++ {
			row = append(row, 0)
		}
		table = append(table, row)
	}

	table[1][1] = 1

	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if j+1 <= n {
				table[i][j+1] += table[i][j]
			}

			if i+1 <= m {
				table[i+1][j] += table[i][j]
			}
		}
	}

	return table[m][n]
}

func main() {
	fmt.Println(GridTraveler(1, 1))   // 1
	fmt.Println(GridTraveler(2, 3))   // 3
	fmt.Println(GridTraveler(3, 2))   // 3
	fmt.Println(GridTraveler(3, 3))   // 6
	fmt.Println(GridTraveler(18, 18)) // 2333606220
}

File: tabulation/python/02_grid_traveler.py
"""Grid traveler.

This script implements a solution to the grid traveler problem using tabulation.

Say that you are a traveler on a 2D grid. You begin in the top-left corner and your goal
is to travel to the bottom-right corner. You may only move down or right.

In how many ways can you travel to the goal on a grid with dimensions m * n?
"""


def grid_traveler(m: int, n: int) -> int:
    """Grid traveler.

    Given the number of lines and columns, return the number of ways the traveler can
    travel to the goal.

    Args:
        m (int): Number of lines.
        n (int): Number of columns.

    Returns:
        int: The number of ways the traveler can travel to the goal.
    """
    table = [[0 for _ in range(n + 1)] for _ in range(m + 1)]
    table[1][1] = 1

    for i in range(m + 1):
        for j in range(n + 1):
            if j + 1 <= n:
                table[i][j + 1] += table[i][j]
            if i + 1 <= m:
                table[i + 1][j] += table[i][j]

    return table[m][n]


if __name__ == "__main__":
    print(grid_traveler(1, 1))  # 1
    print(grid_traveler(2, 3))  # 3
    print(grid_traveler(3, 2))  # 3
    print(grid_traveler(3, 3))  # 6
    print(grid_traveler(18, 18))  # 2333606220
